Added ASG, flagged instances for imdsv2, refactored code

weikequ opened the pull request:
*Issue #, if available:*
Closes following issues
- https://github.com/runfinch/infrastructure/issues/140
- https://github.com/runfinch/infrastructure/issues/138
- https://github.com/runfinch/infrastructure/issues/40
- https://github.com/runfinch/infrastructure/issues/30

*Description of changes:*
- Added autoscaling groups
- Added launch template
- Added scheduling to wind down beta instances after 1 day
- Marked option for IMDSv2 being required for instances
- Instances now auto connect to GitHub repos
- Refactored runner configuration

*Testing done:*
- cdk synth on local
- npm run test
- tested on my own pipeline/accounts


- [ x ] I've reviewed the guidance in CONTRIBUTING.md


#### License Acceptance

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

weikequ commented:
Sorry for the large PR! Wanted to keep downtime of runners as low as possible. Note that the runner configuration is just a placeholder for now as I confirm with @vsiravar and @KevinLiAWS what we need for actual infra.

monirul commented on lib/asg-runner-stack.ts:
```diff
@@ -0,0 +1,173 @@
+ *  - a launch template
+ *  - an auto scaling group
+ */
+export class ASGRunnerStack extends cdk.Stack {
```
Did you verify  if you provision an instance using ASG Runner and it shows up in System manager fleet? Otherwise it might raise security compliance related concerns in PVRE.
Also, try out when it provisions new instances as part of auto scaling.

weikequ replied:
I tested it with my own account. It does show up in System Manager Fleet.

weikequ replied:
It is the same process upon creation of the stack and when it is scaled

monirul commented on lib/asg-runner-stack.ts:
```diff
@@ -0,0 +1,173 @@
+
+    // Create a custom name for this as names for resource groups cannot be repeated
+    const resourceGroup = new resourcegroups.CfnGroup(this, 'Mac' + '-' + props.type.macOSVersion.split('.')[0] + '-' + arch.split('_')[0] + 'HostGroup', {
+      name: 'Mac' + '-' + props.type.macOSVersion.split('.')[0] + '-' + arch.split('_')[0] + 'HostGroup',
```
Same here.

weikequ replied:
Do you mean if the resource group shows up in System Manager? What are you referring to?

weikequ replied:
Refactored to be another variable.

weikequ replied:
@monirul can you check again to see if that is what you mean with these 2 comments? Thank you!

monirul commented on lib/asg-runner-stack.ts:
```diff
@@ -0,0 +1,173 @@
+    }
+
+    // Create a custom name for this as names for resource groups cannot be repeated
+    const resourceGroup = new resourcegroups.CfnGroup(this, 'Mac' + '-' + props.type.macOSVersion.split('.')[0] + '-' + arch.split('_')[0] + 'HostGroup', {
```
Can we add a separate function? Even though it is used in one place, since this one has pretty long formatting. Maybe we can offload that to a separate function. that way it will look cleaner.

weikequ replied:
Sure, we can do that. What do others think?

weikequ replied:
Not too sure what you mean. `resourcegroups` is imported from `aws-resourcegroups`, and `resourceGroup` is the variable I created. Can you clarify? Thanks!

weikequ replied:
Thanks for the comment @pendo324, revised accordingly!

monirul commented on lib/asg-runner-stack.ts:
```diff
@@ -0,0 +1,173 @@
+      launchTemplate: lt
+    })
+
+    if (props.stage === ENVIRONMENT_STAGE.Beta) {
```
Why this is only applicable for Beta. If it is only for testing purpose. Pls add a TODO/comment so that we dont forget about this.

weikequ replied:
This is because we don't actually add the beta instances to the repos, they are spun up to see if we catch any errors, and then spun down after a day (see PR notes and the linked issues).

monirul reviewed (commented):
Added some comments.

pendo324 reviewed (commented):
Left some comments, nothing major though. Thanks for this, looks like a lot of work and research went into it

weikequ commented on config/runner-config.json:
```diff
@@ -0,0 +1,59 @@
+                "macOSVersion": "13.2",
+                "arch": "arm",
+                "repo": "finch",
+                "desiredInstances": 1,
```
See PR comment. Still working with Vishwas and Kevin to settle on the total number of instances we need given the current changes to the build process.

pendo324 replied:
Oh nice, missed your comment. Thanks.

pendo324 commented on lib/asg-runner-stack.ts:
```diff
@@ -0,0 +1,173 @@
+        resources: [ `arn:aws:secretsmanager:${props.env?.region}:${props.env?.account}:secret:${props.type.repo}-runner-reg-key*` ]
+      }))
+    }
+
+    // Create a custom name for this as names for resource groups cannot be repeated
+    const resourceGroup = new resourcegroups.CfnGroup(this, 'Mac' + '-' + props.type.macOSVersion.split('.')[0] + '-' + arch.split('_')[0] + 'HostGroup', {
+      name: 'Mac' + '-' + props.type.macOSVersion.split('.')[0] + '-' + arch.split('_')[0] + 'HostGroup',
```
This is what I was suggesting, not sure if Monirul was mentioning the same thing:

```suggestion

    // Create a custom name for this as names for resource groups cannot be repeated
    const resourceGroupName = `Mac-${props.type.macOSVersion.split('.')[0]}-${arch.split('_')[0]}HostGroup`
    const resourceGroup = new resourcegroups.CfnGroup(this, resourceGroupName, {
      name: resourceGroupName,
```

weikequ replied:
Oh! Got it, okay that makes a lot of sense. Will update, thanks!

weikequ commented on lib/asg-runner-stack.ts:
```diff
@@ -0,0 +1,173 @@
+      throw new Error("Runner environment is undefined!")
+    }
+    
+    const vpc = cdk.aws_ec2.Vpc.fromLookup(this, 'VPC',{isDefault: true});
```
Ran prettier and updated a new version. Thanks!

pendo324 reviewed (approved):
LGTM, pending the correct amount of instances 👍

monirul reviewed (approved):
LGTM.

weikequ commented:
Updated # of instances after confirmation with team, added another modifier to the stack/resource group name for the repo, allowed beta instances to be connected to `finch` and `finch-core` repos as well.
